agent/agentcontainers/api.go

```diff
@@ -1229,14 +1230,38 @@ func (api *API) Close() error {
 	api.logger.Debug(api.ctx, "closing API")
 	api.closed = true
 
-	for _, proc := range api.injectedSubAgentProcs {
-		api.logger.Debug(api.ctx, "canceling subagent process", slog.F("agent_name", proc.agent.Name), slog.F("agent_id", proc.agent.ID))
+	// Stop all running subagent processes and clean up.
+	subAgentIDs := make([]uuid.UUID, 0, len(api.injectedSubAgentProcs))
+	for workspaceFolder, proc := range api.injectedSubAgentProcs {
+		api.logger.Debug(api.ctx, "canceling subagent process",
+			slog.F("agent_name", proc.agent.Name),
+			slog.F("agent_id", proc.agent.ID),
+			slog.F("container_id", proc.containerID),
+			slog.F("workspace_folder", workspaceFolder),
+		)
 		proc.stop()
+		if proc.agent.ID != uuid.Nil {
+			subAgentIDs = append(subAgentIDs, proc.agent.ID)
+		}
 	}
+	api.injectedSubAgentProcs = make(map[string]subAgentProcess)
 
 	api.cancel()    // Interrupt all routines.
 	api.mu.Unlock() // Release lock before waiting for goroutines.
 
+	// Note: We can't use api.ctx here because it's canceled.
+	deleteCtx, deleteCancel := context.WithTimeout(context.Background(), defaultOperationTimeout)
+	defer deleteCancel()
+	for _, id := range subAgentIDs {
+		err := api.subAgentClient.Delete(deleteCtx, id)
+		if err != nil {
+			api.logger.Error(api.ctx, "delete subagent record during shutdown failed",
+				slog.Error(err),
+				slog.F("agent_id", id),
+			)
+		}
+	}
```

Comment:
Suggestion: should we do this in parallel instead?

Reply:
Main argument is to reduce time to shutdown. Currently it is worst case maximum `len(subAgentIDs) * deleteCtx timeout`.

Reply:
Gotcha 👍🏻. It's btw just `deleteCtx` timeout, we give up on deleting the rest if it takes too long.

Reply:
Gotcha. That may be another argument for doing multiple at once - I'd pefer it to be all or nothing.